docs/adr/adr-012-daser-parallelization.md

```diff
@@ -0,0 +1,56 @@
+# ADR #012: DAS'er parallelization
+
+## Changelog
+
+- 2022-9-12: Started
+
+## Authors
+
+@walldiss
+
+## Context
+
+DAS is the process of verifying the availability of
+block data by sampling chunks or shares of those blocks. Package `das` can confirm the availability of block data in the network via the
+Availability interface. High availability of shares is a priority functionality for celestia-node. Its performance could benefit a lot from parallelization optimisation to make it able to fully utilise network bandwidth.
+
+## Previous approach
+
+Shares sampling by its nature is a network bound operation that implies multiple network round-trips.
+Previous implementation of DAS'er was using single thread synchronous approach.
+Meaning there was only one process sequentially performing sampling operation over past headers, that was blocked by awaiting a for response.
+
+## Decision
+
+Usage of multiple coordinated workers running in parallel could drastically improve DAS'er performance by better utilisation of network bandwidth. On the downside, solution brings concurrency complexity.
+
+## Detailed Design
+
+To achieve parallelization DAS'er was split into core components:
+
+1. `Coordinator` holds all information about sampled headers and defines what headers should be sampled next.
+2. `Workers` perform sampling over range of headers and communicate results back to coordinator.
+3. `Subscriber` subscribes to updates of network head. If new head is found it will notify the `Coordinator`. Recent network head blocks will be prioritised for sampling to increase availability of most demanded blocks.
+4. `CheckpointStore` stores / loads `Coordinator` state as a checkpoint to allow seamless resume upon restart. `Coordinator` will store state as checkpoint on exit and will resume sampling from the latest state.
+Also periodically stores checkpoint to storage in case node hard shutdown.
```

Comment:
```suggestion
4. `CheckpointStore` stores/loads` Coordinator` state as a checkpoint to allow seamless resume upon restart. `Coordinator` will store the state as a checkpoint on the exit and will resume sampling from the latest state.
Also periodically stores checkpoint to storage in case node hard shutdown.
```